qiita_db/processing_job.py

```diff
@@ -6,21 +6,295 @@
+                                                              ['depend'])
+                        tmp = Watcher.job_state_map[results['job_state']]
+                        results['job_state'] = tmp
+                        if results['job_state'] == 'completed':
+                            results2 = self._element_extract(item,
+                                                             ['exit_status'],
+                                                             [])
+                            results['exit_status'] = results2['exit_status']
+
+                        # determine if anything has changed since last poll
+                        if results['Job_Id'] in self.processes:
+                            tmp = cmp(self.processes[results['Job_Id']],
+                                      results)
+                            if 0 != tmp:
+                                # metadata for existing job has changed
+                                self.processes[results['Job_Id']] = results
+                                self.queue.put(results)
+                                self._process_dependent_jobs(results)
+                        else:
+                            # metadata for new job inserted
+                            self.processes[results['Job_Id']] = results
+                            self.queue.put(results)
+            else:
+                self.queue.put('QUIT')
+                self.event.set()
+                # don't join(), since we are exiting from the main loop
+
+            sleep(self.polling_value)
+
+    def stop(self):
+        # 'poison pill' to thread/process
+        self.queue.put('QUIT')
+        # setting self.event is a safe way of communicating a boolean
+        # value across processes and threads.
+        # when this event is 'set' by the main line of execution in Qiita,
+        # (or in any other process if need be), Watcher's run loop will
+        # stop and the Watcher process will exit.
+        self.event.set()
+        # Here, it is assumed that we are running this from the main
+        # context. By joining(), we're waiting for the Watcher process to
+        # end before returning from this method.
+        self.join()
+
+
+def launch_local(env_script, start_script, url, job_id, job_dir):
+
+    # launch_local() differs from launch_torque(), as no Watcher() is used.
+    # each launch_local() process will execute the cmd as a child process,
+    # wait, and update the database once cmd has completed.
+    #
+    # As processes are lighter weight than jobs, this should be fine.
+    # This is how the current job model works locally.
+
+    cmd = [start_script, url, job_id, job_dir]
+
+    # When Popen() executes, the shell is not in interactive mode,
+    # so it is not sourcing any of the bash configuration files
+    # We need to source it so the env_script are available
+    cmd = "bash -c '%s; %s'" % (env_script, ' '.join(cmd))
```

Comment:
Would it make sense to add -x to prevent any subsequent commands once a failure has been found?

Reply:
Friendly ping on this.

Reply:
For now, I think it's best to leave that line as we pulled it from the old launcher. However, I think it's a good idea worth testing at a later point. I will add it to the list of TODOs.